### PR TITLE
Add mutual contact requests

### PR DESCRIPTION
**Pull request by cammellos**
this pr adds mutual contact requests and all the related notifications, it's not a breaking change in terms of protocol, and it currently does not do handle syncing or backup, that's going to be built separately.
it also upgrades to golang 1.18

### CODE REVIEW

**Comment by jakubgs** [comment]
I can reproduce it locally:
```
jenkins@linux-01: .../status-go$ make vendor
go mod tidy
go mod vendor
modvendor -copy="**/*.c **/*.h" -v
Error! "/home/jenkins/workspace/status-go_prs_android_PR-2665/pkg/mod/github.com/sirupsen/logrus@v1.4.2" module path does not exist, check $GOPATH/pkg/mod
make: *** [Makefile:334: vendor] Error 1

jenkins@linux-01: .../status-go$ ls -l $GOPATH/pkg/mod/github.com/sirupsen/
total 8
dr-xr-xr-x 6 jenkins jenkins 4096 May 10 10:16 logrus@v1.7.0
dr-xr-xr-x 5 jenkins jenkins 4096 May 10 10:16 logrus@v1.8.1
```
Looks like the issue is this `replace` line in `go.mod` for `logrus@v1.4.2`:
https://github.com/status-im/status-go/blob/137d16137d1de4cee1c7d473a76786f47fa6a9b1/go.mod#L7

File: protocol/messenger.go [hunk 1]
```diff
@@ -3165,10 +3161,19 @@ func (m *Messenger) handleRetrievedMessages(chatWithMessages map[transport.Filte
...
 		Timesource:            m.getTimesource(),
 		AllBookmarks:          make(map[string]*browsers.Bookmark),
 	}
+}
+
+func (m *Messenger) handleRetrievedMessages(chatWithMessages map[transport.Filter][]*types.Message, storeWakuMessages bool) (*MessengerResponse, error) {
+
+	m.handleMessagesMutex.Lock()
+	defer m.handleMessagesMutex.Unlock()
+
```
Comment (Samyoul) [inline protocol/messenger.go:3263]: That's interesting, were you getting concurrency issues?

**Review by Samyoul — APPROVED** [review]
For such a substantial PR I'm a bit sad there is nothing to complain about.

File: protocol/messenger.go (hunk 1 above)
Comment (cammellos) [reply protocol/messenger.go:3263]: It was introduced by @PascalPrecht under my suggestion, when retrieving torrent from community archives (asynchronously), `handleMessages` might be run concurrently, and  that will definitely have some issues I believe.

**Review by richard-ramos — APPROVED** [review]
:100:

**Review by elina2015 — APPROVED** [review]

**Comment by elina2015** [comment]
Tested on Status desktop version: [Build #2137 (May 30, 2022, 9:02:25 AM)](https://ci.status.im/job/status-desktop/job/branches/job/macos/job/master/2137/)

**Pull request merged**